File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if(s.length() == 0){
            return 0;
        }
        set<char> subs;
        int left = 0, right = 0;
        int maxlength = 0;
        while(right < s.length()){
            if (subs.count(s[right]) == 0){
                subs.insert(s[right]);
                maxlength = max(maxlength, (int)subs.size());
                right++;
            } else {
                subs.erase(s[left]);
                left++;
            }
        }
        return maxlength;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        window = {}

        start, max_len = 0, 0
        
        for end in range(len(s)):
            ch = s[end]
            window[ch] = 1 + window.get(ch, 0)
            while window[ch] > 1:
                ch2 = s[start]
                window[ch2] -= 1
                start += 1

            max_len = max(max_len, end - start + 1)
        
        return max_len
